### ISSUE TITLE
Connections to new service pod stuck in conntrack for TCP traffic when externalTrafficPolicy is "local"

### ISSUE THREAD

**Issue by vinay50muddu** [kubernetes/kubernetes#113203]
### What happened?

**Environment:**
kubernetes version: 1.23.6
ExternalTrafficPolicy:  "local"
service type: "Loadbalancer"
kube-proxy mode: iptables
net.netfilter.nf_conntrack_tcp_be_liberal: 1
Metallb controller with BIRD speaker that establishes BGP peering with Datacenter GW(DCGW) and advertises the LoadBalancer Ip's to it.

For a service of type Loadbalancer, the BIRD speaker on a node advertises the Loadbalancer IP to DCGW on a BGP session when a service pod comes up on the same node. When the service pod(serving TCP traffic)) is deleted, the pod comes up on a different node and speaker on this node now advertises the LB IP to the DCGW.  The TCP traffic which was earlier destined to the deleted pod, is now routed to the new pod on this node. This traffic includes:
1. **New TCP SYN connection requests**. After 15s client side disconnect connections that were in status SYN_SENT since connection was not able to be established. Corresponding conntrack entries are removed after about 2 minutes. So this is not an issue.
2. **Existing TCP connections in ESTABLISHED state**. These are now the midstream packets for this endpoint. Sometimes, it is observed that the speaker advertisement is faster than the kube-proxy setting up the IPTABLE rules to the new endpoint. Since no rule has been setup yet this defaults to dropping the packet. When the first packet arrives, a conntrack entry is registered and updates the entry as UNREPLIED. All the subsequent incoming packets with the same tuple that arrive updates the conntrack entry with updated expiry timer. Unless this conntrack entry is deleted or client terminates the connection, the subsequent packets can not be handled as the timer would never expire. A possible solution would be to delete stale conntrack entries before setting up the IPTABLE rules for the given service.

Logs:
MONITORING CONNTRACK
11:34:57.456904501 < no entries
11:34:57.508626152 < connections registered as UNREPLIED appear. At this moment IPTABLE was not updated yet
tcp      6 299 ESTABLISHED src=10.139.17.79 dst=10.143.246.10 sport=34774 dport=443 [UNREPLIED] src=10.143.246.10 dst=10.139.17.79 sport=443 dport=34774 mark=0 use=1
tcp      6 299 ESTABLISHED src=10.139.17.79 dst=10.143.246.10 sport=34824 dport=443 [UNREPLIED] src=10.143.246.10 dst=10.139.17.79 sport=443 dport=34824 mark=0 use=1
tcp      6 299 ESTABLISHED src=10.139.17.79 dst=10.143.246.10 sport=34912 dport=443 [UNREPLIED] src=10.143.246.10 dst=10.139.17.79 sport=443 dport=34912 mark=0 use=1
tcp      6 299 ESTABLISHED src=10.139.17.79 dst=10.143.246.10 sport=34868 dport=443 [UNREPLIED] src=10.143.246.10 dst=10.139.17.79 sport=443 dport=34868 mark=0 use=1
tcp      6 299 ESTABLISHED src=10.139.17.79 dst=10.143.246.10 sport=34834 dport=443 [UNREPLIED] src=10.143.246.10 dst=10.139.17.79 sport=443 dport=34834 mark=0 use=1
tcp      6 299 ESTABLISHED src=10.139.17.79 dst=10.143.246.10 sport=34908 dport=443 [UNREPLIED] src=10.143.246.10 dst=10.139.17.79 sport=443 dport=34908 mark=0 use=1
tcp      6 299 ESTABLISHED src=10.139.17.79 dst=10.143.246.10 sport=34916 dport=443 [UNREPLIED] src=10.143.246.10 dst=10.139.17.79 sport=443 dport=34916 mark=0 use=1
tcp      6 119 SYN_SENT src=10.139.17.79 dst=10.143.246.10 sport=34938 dport=443 [UNREPLIED] src=10.143.246.10 dst=10.139.17.79 sport=443 dport=34938 mark=0 use=1
tcp      6 299 ESTABLISHED src=10.139.17.79 dst=10.143.246.10 sport=34910 dport=443 [UNREPLIED] src=10.143.246.10 dst=10.139.17.79 sport=443 dport=34910 mark=0 use=1
tcp      6 299 ESTABLISHED src=10.139.17.79 dst=10.143.246.10 sport=34790 dport=443 [UNREPLIED] src=10.143.246.10 dst=10.139.17.79 sport=443 dport=34790 mark=0 use=1
tcp      6 299 ESTABLISHED src=10.139.17.79 dst=10.143.246.10 sport=34858 dport=443 [UNREPLIED] src=10.143.246.10 dst=10.139.17.79 sport=443 dport=34858 mark=0 use=1
tcp      6 300 ESTABLISHED src=10.139.17.79 dst=10.143.246.10 sport=34862 dport=443 [UNREPLIED] src=10.143.246.10 dst=10.139.17.79 sport=443 dport=34862 mark=0 use=1

### What did you expect to happen?

Either "RST" is expected to be sent for the midstream packet, or the stale conntrack entries must be removed once IPTABLE rules for the service are configured.

### How can we reproduce it (as minimally and precisely as possible)?

Create a TCP service of type "Loadbalancer" and setting the "externalTrafficPolicy" to "local".
Run the kube-proxy in "iptables" mode.
Run a client application that pumps continues TCP traffic to the Loadbalancer IP on the configured port.
Delete the service pod and the pod comes up in a different node.
Observe the conntrack entries.

### Kubernetes version

<details>
Kubernetes version: 1.23.6
</details>

### Cloud provider

<details>
Vanilla kubernetes
</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
NAME="SLES"
VERSION="15-SP3"
VERSION_ID="15.3"
PRETTY_NAME="SUSE Linux Enterprise Server 15 SP3"
ID="sles"
ID_LIKE="suse"
ANSI_COLOR="0;32"
CPE_NAME="cpe:/o:suse:sles:15:sp3"
DOCUMENTATION_URL="https://documentation.suse.com/"

$ uname -a
Linux worker-pool1-xyz 5.3.18-150300.59.87-default #1 SMP Thu Jul 21 14:31:28 UTC 2022 (cc90276) x86_64 x86_64 x86_64 GNU/Linux


# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

**Comment by aojea** [comment]
> When the service pod(serving TCP traffic)) is deleted, the pod comes up on a different node and speaker on this node now advertises the LB IP to the DCGW. The TCP traffic which was earlier destined to the deleted pod, is now routed to the new pod on this node. This traffic includes:

> 2\. **Existing TCP connections in ESTABLISHED state**. 

Some questions because is no clear the scenario to me:

how can there be existing TCP connections in ESTABLISHED state if the pod was deleted?
Once the pod is deleted it should close the TCP connection, no? that will delete the conntrack entry

**Comment by vinay50muddu** [comment]
Let's say there are 2 nodes NODE-A and NODE-B. The service pod was running on NODE-A earlier. Assuming after we deleted the pod on NODE-A and the pod comes on NODE-B, the speaker on NODE-A sends an update to DCGW to clear the route to NODE-A (as the traffic policy is local, if there are no service pods running on the node, the traffic can't be routed to this node which is NODE-A.). Here i think there is no scope for the pod to send an RST to the client.  When the pod comes up on NODE-B, the speaker advertises the route to DCGW before the IPTABLE rules are setup. The DCGW now redirects the traffic (that was earlier going to pod on NODE-A which is deleted now) to other node (in this case NODE-B) where the service pod is running. So this traffic is a midstream TCP traffic redirected to NODE-B because NODE-A route is not available. Hope this is clear.

**Comment by aojea** [comment]
hmm, this reminds me about https://github.com/kubernetes/kubernetes/issues/101607#issuecomment-867774857

can you check your host kernels have this patch https://github.com/torvalds/linux/commit/e15d4cdf27cb0c1e977270270b2cea12e0955edd ?

**Comment by vinay50muddu** [comment]
@aojea , Thanks for the replies. Yes, I am aware of this. The changes you referred to fixes issue with entries in "SYN_SENT" state. The current issue is reported for conntrack entries with ESTABLISHED state which are marked as UNREPLIED.

**Comment by aojea** [comment]
The connection ESTABLISHED and UNREPLIED seems a problem on the routing change,  breaking the connection shutdown https://netfilter.vger.kernel.narkive.com/911wYpBK/conntrack-and-established-unreplied-connections

@fedepaol I'm no longer familiar with metallb, why metallb should be faster than kube-proxy?

**Comment by fedepaol** [comment]
> The connection ESTABLISHED and UNREPLIED seems a problem on the routing change, breaking the connection shutdown https://netfilter.vger.kernel.narkive.com/911wYpBK/conntrack-and-established-unreplied-connections
> 
> @fedepaol I'm no longer familiar with metallb, why metallb should be faster than kube-proxy?

I am not sure. What MetalLB does is, it listens to endpoints and from each node it checks if there's at least an active end point ready (in case of ETP=local), and updates the BGP advertisements accordingly. I am not sure metallb can be faster than kube-proxy in this scenario, but I am not familiar at all with kube-proxy.

Also, question for @vinay50muddu : are you using a metallb fork? Because the official one is not using BIRD in speakers.

**Comment by vinay50muddu** [comment]
> The connection ESTABLISHED and UNREPLIED seems a problem on the routing change, breaking the connection shutdown https://netfilter.vger.kernel.narkive.com/911wYpBK/conntrack-and-established-unreplied-connections
> 
> @fedepaol I'm no longer familiar with metallb, why metallb should be faster than kube-proxy?

I have gone through the link. As i pointed out in the environment section earlier the kernel flag net.netfilter.nf_conntrack_tcp_be_liberal is set to "1" while the default value for this kernel flag is 0.  Do you see any problem with this?

**Comment by vinay50muddu** [comment]
> > The connection ESTABLISHED and UNREPLIED seems a problem on the routing change, breaking the connection shutdown https://netfilter.vger.kernel.narkive.com/911wYpBK/conntrack-and-established-unreplied-connections
> > @fedepaol I'm no longer familiar with metallb, why metallb should be faster than kube-proxy?
> 
> I am not sure. What MetalLB does is, it listens to endpoints and from each node it checks if there's at least an active end point ready (in case of ETP=local), and updates the BGP advertisements accordingly. I am not sure metallb can be faster than kube-proxy in this scenario, but I am not familiar at all with kube-proxy.
> 
> Also, question for @vinay50muddu : are you using a metallb fork? Because the official one is not using BIRD in speakers.

@fedepaol , we are using metallb old versions where they have the speaker that uses BIRD. Recently they started to use FRR. As far as the race condition between BGP speaker and kube-proxy is concerned, i think it is quite possible. Because since both watch on the service & endpoint events, it is reasonable to think that the advertisements from speaker may precede the iptables update by kube-proxy, though not everytime.

**Comment by aojea** [comment]
> Do you see any problem with this?

This seems that the change of routing is breaking the connection shutdown and leaving this conntrack state, the similar issues we had caused by asymmetric routing were solved by setting tcp_be_liberal to 1 https://www.google.com/search?q=nf_conntrack_tcp_be_liberal+kubernetes

**Comment by vinay50muddu** [comment]
@fedepaol and @aojea , Is there any update/info required from my side? Please let me know.
This flag is already enabled in all the worker nodes. Any suggestions are appreciated.

**Comment by fedepaol** [comment]
@aojea I assume you are talking about metallb's controller. Isn't checking for the endpoint status enough? And also, how can we make this cni agnostic? There are scenarios where we don't even have kube-proxy.

**Comment by aojea** [comment]
> the speaker advertises the route to DCGW before the IPTABLE rules are setup.

honestly, is the only thing that I can think to solve the problem with the scenario explained here, I'm really lost here, I still don't understand why when the pod dies doesn't close the connection and removes the conntrack rule, is like something is odd here ... 

reports like this in the past turned out to be complex network scenarios with non-common configurations (i.e. clients that used TCP with hardcoded source ports)

do you have any bug like this in metallb fede? I would be surprised that if this is a real problem in common scenarios you don't have multiple reports like this in the metallb project...

**Comment by fedepaol** [comment]
> do you have any bug like this in metallb fede? I would be surprised that if this is a real problem in common scenarios you don't have multiple reports like this in the metallb project...

Nope, first time I see something like this

**Comment by thockin** [comment]
any updates here?

**Comment by uablrek** [comment]
**TL;DR;** I was not really able to re-create this problem, but there is a _big difference_ between v1.23.5 and v1.25.3, so @vinay50muddu can you please test if this problem still exist on the latest K8s?

### Test setup

To move a POD from one node to another I used a DaemonSet with `nodeName` set to one node, then applied a manifest with different node. The POD moves fast (within seconds) to the new node.

<details><summary>Manifests</summary>

```yaml
apiVersion: apps/v1
kind: DaemonSet
metadata:
  name: ctraffic-server
spec:
  selector:
    matchLabels:
      app: ctraffic-server
  template:
    metadata:
      labels:
        app: ctraffic-server
    spec:
      nodeName: vm-002
      containers:
      - name: ctraffic-server
        image: registry.nordix.org/cloud-native/ctraffic-server:latest
        imagePullPolicy: IfNotPresent
---
apiVersion: v1
kind: Service
metadata:
  name: ctraffic
spec:
  ipFamilyPolicy: RequireDualStack
  externalTrafficPolicy: Local
  selector:
    app: ctraffic-server
  ports:
  - port: 5003
    name: ctraffic
  type: LoadBalancer
```

To move the POD, apply:
```yaml
apiVersion: apps/v1
kind: DaemonSet
metadata:
  name: ctraffic-server
spec:
  selector:
    matchLabels:
      app: ctraffic-server
  template:
    metadata:
      labels:
        app: ctraffic-server
    spec:
      nodeSelector:
        kubernetes.io/hostname: vm-003
      containers:
      - name: ctraffic-server
        image: registry.nordix.org/cloud-native/ctraffic-server:latest
        imagePullPolicy: IfNotPresent
```

</details>

The loadBalancerIP is set with [assign-lb-ip](https://github.com/Nordix/assign-lb-ip) to 10.0.0.52, and static routes are setup (no bgp routing). To get the sequence right (as I understood the description) the route must be set to the new node _before_ the pod is started on the new node. This is simple to do manually:
```
ip ro replace 10.0.0.52 via 192.168.1.3
kubectl apply -f vm-003.yaml
```

The [ctraffic](https://github.com/Nordix/ctraffic) program is used to generate continuous tcp traffic of 100 KB/S over 10 connections. Here a difference should be noted, a program that just do `connect(2)` will hang 2m which is the Linux default. In ctraffic connects are configured to timeout after 1500ms. So re-connect attempts will be faster than usual.

## Test case

1. Start continuous traffic for 30s, and log statistics
2. Wait 5s
3. Re-route traffic to the new node, then immediately...
4. Mode the POD to the new node
5. Wait for the traffic generation to stop
6. Collect statistics

## Result

Kernel `linux-6.0.0` was used in all test, so these results are _not kernel dependent_.

<img src="https://user-images.githubusercontent.com/37046727/206431062-ca073e96-6c15-4663-8715-cdad5117a113.svg" width="50%" />

**Comment by uablrek** [comment]
For proxy-mode=ipvs the results are better for k8s v1.23.6 with ~6s outage, and about the same for v1.25.3.

**Comment by uablrek** [comment]
I coudn't notice any significant difference with `net.netfilter.nf_conntrack_tcp_be_liberal: 0`.

**Comment by uablrek** [comment]
I added dump of conntrack _after_ the route is changed to the new node but _before_ the POD is moved. For k8s v1.23.6 there are entries similar to the ones described in this issue:
```
tcp      6 299 ESTABLISHED src=192.168.1.201 dst=10.0.0.52 sport=46032 dport=5003 [UNREPLIED] src=10.0.0.52 dst=192.168.1.201 sport=5003 dport=46032 mark=0 use=1
tcp      6 299 ESTABLISHED src=192.168.1.201 dst=10.0.0.52 sport=46092 dport=5003 [UNREPLIED] src=10.0.0.52 dst=192.168.1.201 sport=5003 dport=46092 mark=0 use=1
tcp      6 299 ESTABLISHED src=192.168.1.201 dst=10.0.0.52 sport=46118 dport=5003 [UNREPLIED] src=10.0.0.52 dst=192.168.1.201 sport=5003 dport=46118 mark=0 use=1
tcp      6 299 ESTABLISHED src=192.168.1.201 dst=10.0.0.52 sport=46072 dport=5003 [UNREPLIED] src=10.0.0.52 dst=192.168.1.201 sport=5003 dport=46072 mark=0 use=1
...
```
For v1.25.3 however, there are none.

**Comment by thockin** [comment]
@uablrek Are you confirming this as an actual bug?  I can't quite tell.

**Comment by uablrek** [comment]
@thockin I would say no, it is not a bug, at least not in latest K8s. I _could_ be some race that I was unable to re-create, but timing is very different in v1.25.3 so I would like a re-test on the latest K8s @vinay50muddu

**Comment by adilGhaffarDev** [comment]
@uablrek thank you for your investigation. It is not reproducible in the latest k8s versions, we can close this issue.
